File: Leetcode/0049-group-anagrams/solution.py
# ============================================================
# 0049. Group Anagrams
# https://leetcode.com/problems/group-anagrams/
# Difficulty: Medium
# Tags: Array, Hash Table, String, Sorting
# ============================================================

from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        """
        Optimal Solution (Character Count as Key)
        Approach: Use character frequency tuple as Hash Map key
        Time:  O(n * k) — n strings, each of length k
        Space: O(n * k) — storing all strings in the Hash Map
        """
        # Hash Map: character count tuple → list of original strings
        groups = defaultdict(list)

        for s in strs:
            # Count frequency of each character (a-z)
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1

            # Tuple is hashable — use as key
            # Anagrams produce the same count tuple
            groups[tuple(count)].append(s)

        return list(groups.values())


# ============================================================
# Test Cases
# ============================================================

if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0

    def test(name: str, got: list[list[str]], expected: list[list[str]]):
        global passed, failed
        # Sort inner lists and outer list for comparison
        got_sorted = sorted([sorted(g) for g in got])
        exp_sorted = sorted([sorted(g) for g in expected])
        if got_sorted == exp_sorted:
            print(f"\u2705 PASS: {name}")
            passed += 1
        else:
            print(f"\u274c FAIL: {name}")
            print(f"  Got:      {got}")
            print(f"  Expected: {expected}")
            failed += 1

    # Test 1: Main example
    test("Main example",
         sol.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]),
         [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]])

    # Test 2: Empty string
    test("Empty string",
         sol.groupAnagrams([""]),
         [[""]])

    # Test 3: Single character
    test("Single character",
         sol.groupAnagrams(["a"]),
         [["a"]])

    # Test 4: No anagrams
    test("No anagrams",
         sol.groupAnagrams(["abc", "def", "ghi"]),
         [["abc"], ["def"], ["ghi"]])

    # Test 5: All anagrams
    test("All anagrams",
         sol.groupAnagrams(["abc", "bca", "cab"]),
         [["abc", "bca", "cab"]])

    # Test 6: Duplicate strings
    test("Duplicate strings",
         sol.groupAnagrams(["a", "a"]),
         [["a", "a"]])

    # Test 7: Mixed lengths
    test("Mixed lengths",
         sol.groupAnagrams(["a", "ab", "ba", "abc", "bca"]),
         [["a"], ["ab", "ba"], ["abc", "bca"]])

    # Test 8: Multiple empty strings
    test("Multiple empty strings",
         sol.groupAnagrams(["", ""]),
         [["", ""]])

    # Test 9: Long anagram group
    test("Long anagram group",
         sol.groupAnagrams(["listen", "silent", "enlist", "inlets", "tinsel"]),
         [["listen", "silent", "enlist", "inlets", "tinsel"]])

    # Results
    print(f"\n\U0001f4ca Results: {passed} passed, {failed} failed, {passed + failed} total")

File: Leetcode/0049-group-anagrams/solution.java
import java.util.*;

/**
 * 0049. Group Anagrams
 * https://leetcode.com/problems/group-anagrams/
 * Difficulty: Medium
 * Tags: Array, Hash Table, String, Sorting
 */
class Solution {

    /**
     * Optimal Solution (Character Count as Key)
     * Approach: Use character frequency string as Hash Map key
     * Time:  O(n * k) — n strings, each of length k
     * Space: O(n * k) — storing all strings in the Hash Map
     */
    public List<List<String>> groupAnagrams(String[] strs) {
        // Hash Map: character count string → list of original strings
        Map<String, List<String>> groups = new HashMap<>();

        for (String s : strs) {
            // Count frequency of each character (a-z)
            int[] count = new int[26];
            for (char c : s.toCharArray()) {
                count[c - 'a']++;
            }

            // Convert count to string key with separator
            // e.g., "1#0#0#0#1#0#...#1#0#0#0" for "eat"
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < 26; i++) {
                sb.append(count[i]);
                sb.append('#');
            }
            String key = sb.toString();

            // Add to the group
            groups.computeIfAbsent(key, k -> new ArrayList<>()).add(s);
        }

        // Return all groups
        return new ArrayList<>(groups.values());
    }

    // ============================================================
    // Test Cases
    // ============================================================

    static int passed = 0, failed = 0;

    static void test(String name, List<List<String>> got, List<List<String>> expected) {
        // Sort inner lists and outer list for comparison
        List<String> gotSorted = normalizeGroups(got);
        List<String> expSorted = normalizeGroups(expected);

        if (gotSorted.equals(expSorted)) {
            System.out.printf("✅ PASS: %s%n", name);
            passed++;
        } else {
            System.out.printf("❌ FAIL: %s%n  Got:      %s%n  Expected: %s%n",
                name, got, expected);
            failed++;
        }
    }

    static List<String> normalizeGroups(List<List<String>> groups) {
        List<String> result = new ArrayList<>();
        for (List<String> g : groups) {
            List<String> sorted = new ArrayList<>(g);
            Collections.sort(sorted);
            result.add(String.join(",", sorted));
        }
        Collections.sort(result);
        return result;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Test 1: Main example
        test("Main example",
            sol.groupAnagrams(new String[]{"eat", "tea", "tan", "ate", "nat", "bat"}),
            Arrays.asList(
                Arrays.asList("bat"),
                Arrays.asList("nat", "tan"),
                Arrays.asList("ate", "eat", "tea")));

        // Test 2: Empty string
        test("Empty string",
            sol.groupAnagrams(new String[]{""}),
            Arrays.asList(Arrays.asList("")));

        // Test 3: Single character
        test("Single character",
            sol.groupAnagrams(new String[]{"a"}),
            Arrays.asList(Arrays.asList("a")));

        // Test 4: No anagrams
        test("No anagrams",
            sol.groupAnagrams(new String[]{"abc", "def", "ghi"}),
            Arrays.asList(
                Arrays.asList("abc"),
                Arrays.asList("def"),
                Arrays.asList("ghi")));

        // Test 5: All anagrams
        test("All anagrams",
            sol.groupAnagrams(new String[]{"abc", "bca", "cab"}),
            Arrays.asList(Arrays.asList("abc", "bca", "cab")));

        // Test 6: Duplicate strings
        test("Duplicate strings",
            sol.groupAnagrams(new String[]{"a", "a"}),
            Arrays.asList(Arrays.asList("a", "a")));

        // Test 7: Mixed lengths
        test("Mixed lengths",
            sol.groupAnagrams(new String[]{"a", "ab", "ba", "abc", "bca"}),
            Arrays.asList(
                Arrays.asList("a"),
                Arrays.asList("ab", "ba"),
                Arrays.asList("abc", "bca")));

        // Test 8: Multiple empty strings
        test("Multiple empty strings",
            sol.groupAnagrams(new String[]{"", ""}),
            Arrays.asList(Arrays.asList("", "")));

        // Test 9: Long anagram group
        test("Long anagram group",
            sol.groupAnagrams(new String[]{"listen", "silent", "enlist", "inlets", "tinsel"}),
            Arrays.asList(Arrays.asList("listen", "silent", "enlist", "inlets", "tinsel")));

        // Results
        System.out.printf("%n📊 Results: %d passed, %d failed, %d total%n",
            passed, failed, passed + failed);
    }
}
